[Ajay Anand] iP

Manoharan-Ajay-Anand opened the pull request:
### Duke

> "Live as if you were to die tomorrow. Learn as if you were to live forever." - Mahatma Gandhi


Duke allows you to keep track of your tasks. Advantages of Duke are:

- Keeps track of all tasks
- Saves all tasks after the application is closed
- Super fast interface

All you need to do is:

1. Download it from [here](https://github.com/Manoharan-Ajay-Anand/ip/releases/download/A-Jar/ip.jar)
2. Double click it
3. Add tasks
4. Let it manage for you

Features:

- [ ] Add Todos, Deadlines and Events
- [ ] Mark tasks as done
- [ ] Remove tasks
- [ ] Find tasks by name

And it is **FREE** :grinning:
You can also use it to practise your Java skills. Here is the main method.
```
public static void main(String[] args) {
     Application.launch(args);
}
```

Commit:
Add Gradle support

Manoharan-Ajay-Anand committed:
Greet, Echo, Exit

Manoharan-Ajay-Anand committed:
Add, List

Manoharan-Ajay-Anand committed:
Mark as Done. Added new class Task.

Manoharan-Ajay-Anand committed:
ToDos, Events, Deadlines

Manoharan-Ajay-Anand committed:
Automated Text UI Testing

Manoharan-Ajay-Anand committed:
Handle Errors

Manoharan-Ajay-Anand committed:
Delete. Using List for tasks.

Manoharan-Ajay-Anand committed:
Save to file. Change parsing of commands.

Manoharan-Ajay-Anand committed:
Use Java LocalDate for Deadline and Event

Manoharan-Ajay-Anand committed:
Merge branch 'branch-Level-8'

Manoharan-Ajay-Anand committed:
Add Class UI, Command, Parser, Storage and TaskList

Manoharan-Ajay-Anand committed:
Add packages command, exception, storage, task and ui

Manoharan-Ajay-Anand committed:
Add Test files and methods

Manoharan-Ajay-Anand committed:
Add JavaDoc for all public methods

Manoharan-Ajay-Anand committed:
Follow coding convention

Manoharan-Ajay-Anand committed:
Add find functionality

Manoharan-Ajay-Anand committed:
Merge branch 'branch-A-CodingStandard'

Manoharan-Ajay-Anand committed:
Merge branch 'branch-Level-9'

Manoharan-Ajay-Anand committed:
Set Up Gradle

Manoharan-Ajay-Anand committed:
Merge branch 'add-gradle-support'

Manoharan-Ajay-Anand committed:
Add Gradle support. Change mainClassName in build.gradle.

Manoharan-Ajay-Anand committed:
Add checkstyle. Refactor code to convention. Add abstract class Command with subclasses and execute method

Manoharan-Ajay-Anand committed:
Add JavaFX for application. Update Command classes to work with new UIPane.

Manoharan-Ajay-Anand committed:
Update error messages for Parser

Manoharan-Ajay-Anand committed:
Update JavaDoc for AddCommand

Manoharan-Ajay-Anand committed:
Rename showError to showErrorMessage in UiPane

Manoharan-Ajay-Anand committed:
Update JavaDoc for class members of UIPane

wanyu-l commented on src/main/java/duke/storage/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+     * @param serialNo The serial number of the task to remove.
+     * @return The task that was removed.
+     */
+    public Task remove(int serialNo) throws DukeException {
+        if (tasks.isEmpty()) {
+            throw new DukeException("Sorry Boss, there is no task to remove.");
+        } else if (serialNo < 1 || serialNo > tasks.size()) {
+            throw new DukeException("Sorry Boss, please provide the correct serial no.");
+        }
+        return this.tasks.remove(serialNo - 1);
+    }
```
Just a minor suggestion, considering "delete" would be the input for this method to be called, maybe using delete as method name and printed output would be better?

wanyu-l reviewed (commented):
Looks really good! I find your code really well structured and nicely written.
If I had to nitpick, it would be the "remove" method in TaskList which is commented below.

e0543978 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,33 @@
+import duke.ui.UiPane;
+
+public class AddCommand extends Command {
+    /* The task to add when executing */
```
Perhaps you could add this comment as a Javadoc?

e0543978 reviewed (commented):
I think your code is very tidy and easy to read. Perhaps one way to better your already polished work is to consider adding Javadocs for the various Command classes and its methods. All the best!

wz27 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,90 @@
+        }
+    }
+
+    /**
+     * Starts the Duke application with the UI. Writes to the data file after being closed by the user.
+     */
```
Should there be @param in this Javadoc comment?

wz27 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @Override
+    public void start(Stage stage) {
```
Should there be a Javadoc header comment for this public method?

wz27 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,90 @@
+ * Represents the main Duke application.
+ */
+public class Duke extends Application {
+    /**
+     * The file path for storing task data
+     */
+    private static final String filePath = "duke.txt";
+    /**
+     * The storage for the application
+     */
+    private Storage storage;
+    /**
+     * The task list for the application
+     */
+    private TaskList taskList;
+    /**
+     * The UI for the application
+     */
+    private UiPane uiPane;
+    /**
+     * The parser for the commands.
+     */
+    private Parser parser;
```
Should Javadoc of class members be specified on a single line?

wz27 commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,28 @@
+import duke.ui.UiPane;
+
+public class DeleteCommand extends Command {
+    private final int serialNo;
```
Should constant names be all uppercase using underscore to separate words?

wz27 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,66 @@
+                return new DoneCommand(serialNo);
+            } else if (cmd.startsWith("deadline ")) {
+                String[] splitText = cmd.substring(9).split(" /by ");
+                Task task = new Deadline(splitText[0].trim(), LocalDate.parse(splitText[1].trim()), false);
```
Could this expression perhaps be less complicated?

wz27 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,66 @@
+     * @return The command after parsing the command string.
+     * @throws DukeException when the given command string cannot be parsed.
+     */
+    public Command parse(String cmd) throws DukeException {
```
Could this long method be perhaps refactored into less than 30 lines of code?

wz27 reviewed (commented):
Looks good! Some very minor code quality changes requested.

Manoharan-Ajay-Anand committed:
Use RegEx to match commands in parse() method in Parser

Using RegEx reduces the text checking and processing done by us. It also guarantees that the command entered matches the proper format leading to fewer bugs.

Manoharan-Ajay-Anand committed:
Enable assertions in gradle run task

Enables Java assertions when running the app.

Use Java assert in parse() method in Parser

Command strings should have matched with the RegEx within the if-else blocks. Assumptions of the string length > n must hold true so that .substring(n) does not throw an Exception. Assumptions of the length of the string array after calling split() must also hold true so that array accesses does not throw an Exception.

Manoharan-Ajay-Anand committed:
Rename remove method in TaskList to delete

Delete command is used in App and other classes use the term delete. Renaming to delete ensures consistency.

Broke up parse method in Parser

parse method exceeded 30 LOC previously. Broke it up into several private methods that abstract parsing task commands and special commands.

Manoharan-Ajay-Anand committed:
Merge pull request #1 from Manoharan-Ajay-Anand/branch-A-Assertions

Java assertions

Manoharan-Ajay-Anand committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/parser/Parser.java

Manoharan-Ajay-Anand committed:
Merge pull request #2 from Manoharan-Ajay-Anand/branch-A-CodeQuality

Rename remove method in TaskList to delete

Manoharan-Ajay-Anand committed:
C-Update

Manoharan-Ajay-Anand committed:
Update RegEx in Parser parseUpdateCommand()

Easier RegEx to understand to match /date in update command. After the date, it must either be the end of string or a space followed by any number of characters.

Manoharan-Ajay-Anand committed:
Add CommandParser class

The CommandParser is capable of parsing string input to various commands. The CommandParser can be easily extended and registered with other CommandParser instances to handle new commands in the event of new requirements.

Manoharan-Ajay-Anand committed:
Correct JavDoc for Duke class

Correct JavaDoc for Duke class members to single-line comments instead of multi-line comments.

Manoharan-Ajay-Anand committed:
Update User Guide

Added all the available commands and their usage to the UserGuide. Added screenshot of the UI.

Manoharan-Ajay-Anand committed:
Update extension of UI image

Updated extension Ui image from Ui.PNG to Ui.png

Manoharan-Ajay-Anand committed:
Update JavaDoc and remove unused import for CheckStyle

Update the JavaDoc for UpdateCommand and GlobalParser. Remove unused import in DeleteParser.

Manoharan-Ajay-Anand committed:
Add Launcher class

Add Launcher class with main method to launch Duke application as creating jar file causes JavaFX related errors.

Manoharan-Ajay-Anand committed:
Remove javafx block from build.gradle

Removed javafx block from build.gradle as only Windows native binaries were being packaged into the jar file and therefore, jar file was unable to run on other platforms.
